File: structural-design-pattern/composite-pattern/composite.go
// Composite Pattern
// =================
// Category: Structural Design Pattern
//
// Intent:
//   Compose objects into tree structures to represent part-whole hierarchies.
//   Clients treat individual objects (leaves) and compositions (branches) uniformly.
//
// When to use:
//   - Tree/hierarchical structures (files/folders, org charts, UI components)
//   - When clients should treat single objects and groups the same way
//   - When operations should propagate recursively through the tree
//
// Key Participants:
//   - Component (interface): Common interface for leaves and composites
//   - Leaf: End node (File)
//   - Composite: Node with children (Directory)

package main

import (
	"fmt"
	"strings"
)

// ---------------------------------------------------------------------------
// Component — common interface
// ---------------------------------------------------------------------------

type FileSystemEntry interface {
	Name() string
	Size() int
	Display(indent int) string
	Search(keyword string) []string
}

// ---------------------------------------------------------------------------
// Leaf — a file (no children)
// ---------------------------------------------------------------------------

type File struct {
	name      string
	sizeBytes int
}

func NewFile(name string, size int) *File {
	return &File{name: name, sizeBytes: size}
}

func (f *File) Name() string { return f.name }
func (f *File) Size() int    { return f.sizeBytes }

func (f *File) Display(indent int) string {
	prefix := strings.Repeat("  ", indent)
	return fmt.Sprintf("%sF %s (%d bytes)", prefix, f.name, f.sizeBytes)
}

func (f *File) Search(keyword string) []string {
	if strings.Contains(strings.ToLower(f.name), strings.ToLower(keyword)) {
		return []string{f.name}
	}
	return nil
}

// ---------------------------------------------------------------------------
// Composite — a directory (has children)
// ---------------------------------------------------------------------------

type Directory struct {
	name     string
	children []FileSystemEntry
}

func NewDirectory(name string) *Directory {
	return &Directory{name: name}
}

func (d *Directory) Name() string { return d.name }

func (d *Directory) Add(entries ...FileSystemEntry) *Directory {
	d.children = append(d.children, entries...)
	return d
}

// Size recursively sums all children — uniform treatment
func (d *Directory) Size() int {
	total := 0
	for _, child := range d.children {
		total += child.Size()
	}
	return total
}

func (d *Directory) Display(indent int) string {
	prefix := strings.Repeat("  ", indent)
	lines := []string{fmt.Sprintf("%sD %s/ (%d bytes)", prefix, d.name, d.Size())}
	for _, child := range d.children {
		lines = append(lines, child.Display(indent+1))
	}
	return strings.Join(lines, "\n")
}

func (d *Directory) Search(keyword string) []string {
	var results []string
	if strings.Contains(strings.ToLower(d.name), strings.ToLower(keyword)) {
		results = append(results, d.name+"/")
	}
	for _, child := range d.children {
		results = append(results, child.Search(keyword)...)
	}
	return results
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
func main() {
	src := NewDirectory("src")
	src.Add(NewFile("main.go", 2048), NewFile("utils.go", 1024), NewFile("config.go", 512))

	tests := NewDirectory("tests")
	tests.Add(NewFile("test_main.go", 1536), NewFile("test_utils.go", 768))

	docs := NewDirectory("docs")
	docs.Add(NewFile("README.md", 4096), NewFile("CHANGELOG.md", 2048))

	root := NewDirectory("project")
	root.Add(src, tests, docs, NewFile(".gitignore", 128))

	fmt.Println(root.Display(0))
	fmt.Printf("\nTotal project size: %d bytes\n", root.Size())
	fmt.Printf("Source code size: %d bytes\n", src.Size())
	fmt.Printf("\nSearch for 'test': %v\n", root.Search("test"))
	fmt.Printf("Search for 'main': %v\n", root.Search("main"))
}

File: structural-design-pattern/composite-pattern/composite.rs
//! Composite Pattern
//! =================
//! Category: Structural Design Pattern
//!
//! Intent:
//!   Compose objects into tree structures to represent part-whole hierarchies.
//!   Clients treat individual objects (leaves) and compositions (branches) uniformly.
//!
//! When to use:
//!   - Tree/hierarchical structures (files/folders, org charts, UI components)
//!   - When clients should treat single objects and groups the same way
//!
//! Key Participants:
//!   - Component (enum or trait): Common interface
//!   - Leaf: End node
//!   - Composite: Node with children
//!
//! Rust approach:
//!   We use an enum rather than trait objects. Enums are idiomatic for closed
//!   hierarchies and avoid the complexity of trait object lifetimes.

// ---------------------------------------------------------------------------
// Component — an enum that can be either a File (leaf) or Directory (composite)
// ---------------------------------------------------------------------------
enum FileSystemEntry {
    File {
        name: String,
        size_bytes: u64,
    },
    Directory {
        name: String,
        children: Vec<FileSystemEntry>,
    },
}

impl FileSystemEntry {
    fn name(&self) -> &str {
        match self {
            FileSystemEntry::File { name, .. } => name,
            FileSystemEntry::Directory { name, .. } => name,
        }
    }

    /// Returns total size in bytes — recursively sums for directories.
    fn size(&self) -> u64 {
        match self {
            FileSystemEntry::File { size_bytes, .. } => *size_bytes,
            FileSystemEntry::Directory { children, .. } => {
                children.iter().map(|c| c.size()).sum()
            }
        }
    }

    /// Tree-like string representation.
    fn display(&self, indent: usize) -> String {
        let prefix = "  ".repeat(indent);
        match self {
            FileSystemEntry::File { name, size_bytes } => {
                format!("{}F {} ({} bytes)", prefix, name, size_bytes)
            }
            FileSystemEntry::Directory { name, children } => {
                let mut lines = vec![format!("{}D {}/ ({} bytes)", prefix, name, self.size())];
                for child in children {
                    lines.push(child.display(indent + 1));
                }
                lines.join("\n")
            }
        }
    }

    /// Search for entries whose name contains the keyword.
    fn search(&self, keyword: &str) -> Vec<String> {
        let kw = keyword.to_lowercase();
        match self {
            FileSystemEntry::File { name, .. } => {
                if name.to_lowercase().contains(&kw) {
                    vec![name.clone()]
                } else {
                    vec![]
                }
            }
            FileSystemEntry::Directory { name, children } => {
                let mut results = Vec::new();
                if name.to_lowercase().contains(&kw) {
                    results.push(format!("{}/", name));
                }
                for child in children {
                    results.extend(child.search(keyword));
                }
                results
            }
        }
    }
}

// Convenience constructors
fn file(name: &str, size: u64) -> FileSystemEntry {
    FileSystemEntry::File {
        name: name.to_string(),
        size_bytes: size,
    }
}

fn dir(name: &str, children: Vec<FileSystemEntry>) -> FileSystemEntry {
    FileSystemEntry::Directory {
        name: name.to_string(),
        children,
    }
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
fn main() {
    let root = dir("project", vec![
        dir("src", vec![
            file("main.rs", 2048),
            file("utils.rs", 1024),
            file("config.rs", 512),
        ]),
        dir("tests", vec![
            file("test_main.rs", 1536),
            file("test_utils.rs", 768),
        ]),
        dir("docs", vec![
            file("README.md", 4096),
            file("CHANGELOG.md", 2048),
        ]),
        file(".gitignore", 128),
    ]);

    println!("{}", root.display(0));
    println!("\nTotal project size: {} bytes", root.size());
    println!("\nSearch for 'test': {:?}", root.search("test"));
    println!("Search for 'main': {:?}", root.search("main"));
}
